File: code/generate_strategies/IR(pseudocode)/cpp/cpp/function_bitarray_count.cpp
#include <string>
#include <stdexcept>
#include <boost/dynamic_bitset.hpp>

class CountBitsFunction {
    
public:
    int count_bits(const std::string& bits, int value) {
        if (value != 0 && value != 1) {
            throw std::invalid_argument("Value must be 0 or 1.");
        }

        if (bits.empty() || bits.find_first_not_of("01") != std::string::npos) {
            throw std::invalid_argument("Input must be a string of '0' and '1'.");
        }

        boost::dynamic_bitset<> bitset(bits);
        return value == 1 ? bitset.count() : bitset.size() - bitset.count();
    }
};

File: data/python/src/function_bitarray_count.py
from bitarray import bitarray


class CountBitsFunction:
    def count_bits(self, bits, value=1):
        if not isinstance(bits, str):
            raise TypeError("Input must be a string of '0' and '1'.")
        if value not in (0, 1):
            raise ValueError("Value must be either 0 or 1.")

        ba = bitarray(bits)
        return ba.count(value)
